OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,757 @@
+    if (vulkan_layer_producer_ != nullptr) {
+      vulkan_layer_producer_->SetCaptureStatusListener(this);
+    }
+  }
+
+  // Sets the maximal depth of debug markers within one command buffer. If set to 0, command buffers
+  // won't be limited by depth.
+  void SetMaxLocalMarkerDepthPerCommandBuffer(uint32_t max_local_marker_depth_per_command_buffer) {
+    max_local_marker_depth_per_command_buffer_ = max_local_marker_depth_per_command_buffer;
+  }
+
+  void TrackCommandBuffers(VkDevice device, VkCommandPool pool,
+                           const VkCommandBuffer* command_buffers, uint32_t count) {
+    absl::WriterMutexLock lock(&mutex_);
+    if (!pool_to_command_buffers_.contains(pool)) {
+      pool_to_command_buffers_[pool] = {};
+    }
+    absl::flat_hash_set<VkCommandBuffer>& associated_cbs = pool_to_command_buffers_.at(pool);
+    for (uint32_t i = 0; i < count; ++i) {
+      VkCommandBuffer cb = command_buffers[i];
+      associated_cbs.insert(cb);
+      command_buffer_to_device_[cb] = device;
+    }
+  }
+
+  void UntrackCommandBuffers(VkDevice device, VkCommandPool pool,
+                             const VkCommandBuffer* command_buffers, uint32_t count) {
+    absl::WriterMutexLock lock(&mutex_);
+    CHECK(pool_to_command_buffers_.contains(pool));
+    absl::flat_hash_set<VkCommandBuffer>& associated_command_buffers =
+        pool_to_command_buffers_.at(pool);
+    for (uint32_t i = 0; i < count; ++i) {
+      VkCommandBuffer command_buffer = command_buffers[i];
+      associated_command_buffers.erase(command_buffer);
+      CHECK(command_buffer_to_device_.contains(command_buffer));
+      CHECK(command_buffer_to_device_.at(command_buffer) == device);
+      command_buffer_to_device_.erase(command_buffer);
+    }
+    if (associated_command_buffers.empty()) {
+      pool_to_command_buffers_.erase(pool);
+    }
+  }
+
+  void MarkCommandBufferBegin(VkCommandBuffer command_buffer) {
+    // Even when we are not capturing we create state for this command buffer to allow the
+    // debug marker tracking.
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(!command_buffer_to_state_.contains(command_buffer));
+      command_buffer_to_state_[command_buffer] = {};
+    }
+    if (!IsCapturing()) {
+      return;
+    }
+
+    uint32_t slot_index = RecordTimestamp(command_buffer, VK_PIPELINE_STAGE_TOP_OF_PIPE_BIT);
+    {
+      absl::WriterMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_state_.contains(command_buffer));
+      command_buffer_to_state_.at(command_buffer).command_buffer_begin_slot_index =
+          std::make_optional(slot_index);
+    }
+  }
+
+  void MarkCommandBufferEnd(VkCommandBuffer command_buffer) {
+    if (!IsCapturing()) {
+      return;
+    }
+
+    uint32_t slot_index = RecordTimestamp(command_buffer, VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT);
+
+    {
+      absl::ReaderMutexLock lock(&mutex_);
+      CHECK(command_buffer_to_state_.contains(command_buffer));
```

Comment:
Maybe add a comment why we are guaranteed to have the command buffer in command_buffer_to_state_. My understanding is that this is because both MarkCommandBufferBegin/End are called from within the same submit so when ...End is called, we know there has been a corresponding ...Begin call earlier.